compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -1465,6 +1465,21 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
         case tree => tree
   }
 
+  def specializeEq(tree: Tree): Tree =
+    tree match
+      case Apply(sel @ Select(arg1, opName), arg2 :: Nil)
+      if sel.symbol == defn.Any_== || sel.symbol == defn.Any_!= =>
+        defn.ScalaValueClasses().find { cls =>
+          arg1.tpe.derivesFrom(cls) && arg2.tpe.derivesFrom(cls)
+        } match {
+          case Some(cls) =>
+            val newOp = cls.requiredMethod(opName, List(cls.typeRef))
+            arg1.select(newOp).withSpan(sel.span).appliedTo(arg2).withSpan(tree.span)
+          case None => tree
+        }
+      case _ =>
+        tree
```

Comment:
What are the criteria for deciding what methods should be specialized? Is it possible to generalize the specialization?

Reply:
The criteria is that it is an `==` or `!=` on `Any` 
```
sel.symbol == defn.Any_== || sel.symbol == defn.Any_!=
```
and that the types of the compared values are the same value class
```
defn.ScalaValueClasses().find { cls =>
          arg1.tpe.derivesFrom(cls) && arg2.tpe.derivesFrom(cls)
```

Reply:
We might generalize it for different classes. For example Short == Int.

The other potential generalization is on `Numeric` ops but that is much more complex.

Reply:
I am thinking why even specialize `==` and `!=`. This seems to related to the discussion about whether inlining should re-resolve selections. Such specialization helps very little in that perspective, I'm thinking whether it's worthwhile to complicate the compiler.

Reply:
No, in this case it is about the partial evaluation of the operation. If we have `(2: Any) == (3: Any)` we know that this will end calling `2 == 3` and return that result.

All the testing frameworks will benefit from this as the variants of `assertEquals` will generate this code.  That is already a large enough use case.